dss/operations/secrets.py

```diff
@@ -291,3 +293,163 @@ def del_secret(argv: typing.List[str], args: argparse.Namespace):
+        self.type = ['service_account']
+        self.auth_uri = ['https://auth.data.humancellatlas.org/oauth/authorize',
+                         'https://auth.dev.data.humancellatlas.org/oauth/authorize']
+        self.token_uri = ['https://auth.data.humancellatlas.org/oauth/token',
+                          'https://auth.dev.data.humancellatlas.org/oauth/token']
+
+        self.app_secret_name = os.environ['GOOGLE_APPLICATION_SECRETS_SECRETS_NAME']
+        self.gcp_cred_secret_name = os.environ['GOOGLE_APPLICATION_CREDENTIALS_SECRETS_NAME']
+        self.app_secret = self.fetch_secret(self.app_secret_name)
+        self.gcp_cred_secret = self.fetch_secret(self.gcp_cred_secret_name)
+
+    def run_cmd(self, cmd, cwd=os.getcwd(), shell=True):
+        """Run a command and return stdout"""
+        p = subprocess.Popen(cmd,
+                             shell=shell,
+                             stdout=subprocess.PIPE,
+                             stderr=subprocess.PIPE,
+                             cwd=cwd,
+                             env=self.stage_env)
+        stdout, stderr = p.communicate()
+        if p.returncode != 0:
+            raise RuntimeError(f'While checking secrets, an error occured:\n'
+                               f'stdout: {stdout}\n\n'
+                               f'stderr: {stderr}\n')
+        return stdout.decode('utf-8')
+
+    def get_stage_env(self, env_file):
+        """Source the environment file for this stage, then export the current environment as a dict"""
+        dump = 'python -c "import os, json; print(json.dumps(dict(os.environ)))"'
+        cmd = ['bash', '-c', f'source {env_file} && {dump}']
+        return json.loads(self.run_cmd(cmd, shell=False))
+
+    def fetch_secret(self, secret_name):
+        """Fetch the value of a secret from the AWS secrets manager"""
+        try:
+            secret = fetch_secret_safely(secret_name)
+        except RuntimeError:
+            self.missing_secrets.append(secret_name)
+            return
+        except json.decoder.JSONDecodeError:
+            self.missing_secrets.append(secret_name)
+            return
+        if not (('installed' not in secret) or ('client_email' not in secret)) and (self.stage in self.stages):
+            self.malformed_secrets.append(secret_name)
+            return
+        return secret
+
+    def fetch_terraform_output(self, output_name, output_infra_dir):
+        """
+        Use the makefiles in the data-store repository to create terraform resource files and use these to
+        check environment variables. See: https://www.terraform.io/docs/commands/output.html
+        """
+        dss_root_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..'))
+        output_infra_dir = os.path.abspath(os.path.join(dss_root_dir, 'infra', output_infra_dir))
+
+        # populate infra vars for the current stage using terraform
+        self.run_cmd(cmd=f'make -C infra', cwd=dss_root_dir)
```

Comment:
if `make -C infra` gets held up on waiting for a `yes` to apply changes what happens? I've never seen it happen with the current version, just playing devils advocate :)

Reply:
I've noticed that this check_secrets script takes a long time, it might be due to it attempting to build out every single terraform file, if all we need is `gcp_service_account` we could use `make -C infra COMPONENT=gcp_service_account` which would get us exactly what we need, and faster CI deployments :)

Reply:
The `run_command()` function will wait for the command to finish, and will raise a RuntimeError if it does not complete successfully. If `make -C infra` were held up, the CI system would would sit and wait for the check secrets test to finish; if the `make -C infra` command finished with a failure, the tests would also fail. If the command never finished, the test would time out on Allspark (after 30 minutes, I think).

Reply:
re adding `COMPONENT=gcp_service_account` doesn't seem to give much of a speedup. Changing the make command to `make -C infra COMPONENT=gcp_service_account` takes almost a minute:

```
$ time GOOGLE_APPLICATION_CREDENTIALS="${PWD}/gcp-credentials.json" ./scripts/dss-ops.py secrets check
Now checking the secrets for dev...
Secret check complete for dev.

real	0m54.767s
user	0m12.624s
sys	0m3.506s
```

while the original script using `make -C infra` (no component flag) takes ~50 seconds:

```
$ time GOOGLE_APPLICATION_CREDENTIALS="${PWD}/gcp-credentials.json" ./scripts/dss-ops.py secrets check
Now checking the secrets for dev...
Secret check complete for dev.

real	0m48.812s
user	0m12.521s
sys	0m3.221s
```

Passing an env var to subprocess this way may not actually be setting the env var, though, so I need to check whether the command is really exporting the component variable.